pkg/proxy/log.go

```diff
@@ -50,11 +75,29 @@ func SetMetadata(md []string) {
 	metadata = md
 }
 
+func logString(outStr string, retry bool) {
+	_, err := logBuf.WriteString(outStr)
+	if err != nil {
+		if retry {
+			log.WithFields(log.Fields{
+				fieldPath: logPath,
+			}).WithError(err).Warning("Error encountered while writing to access log, retrying once...")
+
+			CloseLogfile()
+			OpenLogfile(logPath)
+
+			// retry once
+			logString(outStr, false)
+		}
+	}
+}
+
 // Log logs a record to the logfile and flushes the buffer
 func Log(l *LogRecord, typ FlowType, verdict FlowVerdict, code int) {
-	if logBuf == nil {
-		return
-	}
+	// Lock while writing access log so we serialize writes as we may have
+	// to reopen the logfile and parallel writes could fail because of that
+	logMutex.Lock()
+	defer logMutex.Unlock()
```

Comment:
Can't we use logrus to write directly to the file instead?

something like this (untested code):
```
import (
	"bufio"
	"encoding/json"
	"fmt"
	"os"

	"github.com/Sirupsen/logrus"
)

var (
	log      = logrus.New()
	logBuf   *bufio.Writer
	metadata []string
)

// OpenLogfile opens a file for logging
func OpenLogfile(lf string) error {
	logFile, err := os.OpenFile(lf, os.O_CREATE|os.O_APPEND|os.O_WRONLY, 0666)
	if err != nil {
		return err
	}
	log.Out = logFile
	return nil
}
```

Reply:
What's the advantage of that? Logrus does not even attempt to reopen the logfile.

Reply:
I see, I thought it would be simpler to leave the file writes to logrus